cc/mcts_node.h

```diff
@@ -0,0 +1,146 @@
+    kSelected,
+
+    // Node has been expanded and is no longer a leaf.
+    kExpanded,
+  };
+
+  static bool CmpN(const EdgeStats& a, const EdgeStats& b) { return a.N < b.N; }
+  static bool CmpW(const EdgeStats& a, const EdgeStats& b) { return a.W < b.W; }
+  static bool CmpP(const EdgeStats& a, const EdgeStats& b) { return a.P < b.P; }
+
+  // Constructor for root node in the tree.
+  MctsNode(EdgeStats* stats, const Position& position);
+
+  // Constructor for child nodes.
+  MctsNode(MctsNode* parent, Coord move);
+
+  float N() const { return stats->N; }
+  float W() const { return stats->W; }
+  float P() const { return stats->P; }
+  float original_P() const { return stats->original_P; }
+  float Q() const { return W() / (1 + N()); }
+  float Q_perspective() const {
+    return position.to_play() == Color::kBlack ? Q() : -Q();
+  }
+
+  float child_N(int i) const { return edges[i].N; }
+  float child_W(int i) const { return edges[i].W; }
+  float child_P(int i) const { return edges[i].P; }
+  float original_child_P(int i) const { return edges[i].original_P; }
+  float child_Q(int i) const { return child_W(i) / (1 + child_N(i)); }
+  float child_U(int i) const {
+    return kPuct * std::sqrt(1.0f + child_N(i)) * child_P(i) / (1 + child_N(i));
+  }
+
+  std::string Describe() const;
+  std::string MostVisitedPath() const;
+
+  void InjectNoise(const std::array<float, kNumMoves>& noise);
+
+  // Selects the next leaf node for inference.
+  // If inference is being batched and SelectLeaf chooses a node that has
+  // already been added to the batch (IncorporateResults has not yet been
+  // called), then SelectLeaf will return null.
+  MctsNode* SelectLeaf();
+
+  void IncorporateResults(absl::Span<const float> move_probabilities,
+                          float value, MctsNode* up_to);
+
+  void IncorporateEndGameResult(float value, MctsNode* up_to);
+
+  void BackupValue(float value, MctsNode* up_to);
+
+  void AddVirtualLoss(MctsNode* up_to);
+
+  void RevertVirtualLoss(MctsNode* up_to);
+
+  // Remove all children from the node except c.
+  void PruneChildren(Coord c);
+
+  // TODO(tommadams): Validate returning by value has the same performance as
+  // passing a pointer to the output array.
+  std::array<float, kNumMoves> CalculateChildActionScore() const;
+
+  MctsNode* MaybeAddChild(Coord c);
+
+  // Parent node.
+  MctsNode* parent;
+
+  // Stats for the edge from parent to this.
+  EdgeStats* stats;
+
+  // Move that led to this position.
+  Coord move;
+
+  std::array<EdgeStats, kNumMoves> edges;
+
+  // TODO(tommadams): a more compact representation.
+  std::array<float, kNumMoves> illegal_moves;
+
+  // Map from move to resulting MctsNode.
+  // TODO(tommadams): use a better containiner.
+  std::unordered_map<int, std::unique_ptr<MctsNode>> children;
+
+  State state = State::kCollapsed;
+
+  // Current board position.
+  Position position;
+
+  // Input features for this position.
+  DualNet::BoardFeatures features;
```

Comment:
The C++ handles the features slightly differently to the Python code. The C++ Position class doesn't keep a history of the board state and so is not able to calculate the input features in isolation. I did it that way because I tried to keep all the MCTS & ML code out of the Position class.

Instead, the MCTS node holds the features and DualNet::UpdateFeatures calculates features for a child node using the parent's features and the child's board state.

When the MctsPlayer plays a move, it does prune all the root's unplayed children, which frees up most of the features arrays. We do keep around the features for the full history of moves played, but this is only around 12MB (19*19*17*500*4) so I don't think we need to worry about it too much.

Reply:
Ah, I see now - you push the history back 2 slots and then prepend the new state, to incrementally compute the feature planes from the previous position.